Sources/Controllers/MessagesViewController.swift

```diff
@@ -251,49 +251,77 @@ extension MessagesViewController: UICollectionViewDataSource {
     }
 
     func removeKeyboardObservers() {
         NotificationCenter.default.removeObserver(self, name: .UIKeyboardWillChangeFrame, object: nil)
+        NotificationCenter.default.removeObserver(self, name: .UIKeyboardDidChangeFrame, object: nil)
         NotificationCenter.default.removeObserver(self, name: .UITextViewTextDidBeginEditing, object: messageInputBar.inputTextView)
     }
 
     @objc
-    func handleTextViewDidBeginEditing(_ notification: Notification) {
+    private func handleTextViewDidBeginEditing(_ notification: Notification) {
         if scrollsToBottomOnKeybordBeginsEditing {
             messagesCollectionView.scrollToBottom(animated: true)
         }
     }
 
+    /// iPad supports undocked and split keyboard modes.
+    /// https://support.apple.com/en-us/HT207521
+    /// When you drag the keyboard in those states across the screen
+    /// `UIKeyboardFrameEndUserInfoKey` in `UIKeyboardWillChangeFrame` notification has zero height.
+    /// This leads to wierd animations and an incorrect state of `messagesCollectionView` bottom inset.
+    /// Luckly we always receive the keyboard end frame in `UIKeyboardDidChangeFrame` notification.
+    /// We can use it to animate the UI to the correct state.
+    ///
+    /// So, if the keyboard end frame in `UIKeyboardWillChangeFrame` has zero height,
+    /// we ignore it and calculate layout adjustments in `UIKeyboardDidChangeFrame` handler instead.
     @objc
-    func handleKeyboardDidChangeState(_ notification: Notification) {
-
+    private func handleKeyboardDidChangeState(_ notification: Notification) {
         guard let keyboardEndFrame = notification.userInfo?[UIKeyboardFrameEndUserInfoKey] as? CGRect else { return }
 
-        if (keyboardEndFrame.origin.y + keyboardEndFrame.size.height) > UIScreen.main.bounds.height {
-            // Hardware keyboard is found
-            let bottomInset = view.frame.size.height - keyboardEndFrame.origin.y - iPhoneXBottomInset
-            messagesCollectionView.contentInset.bottom = bottomInset
-            messagesCollectionView.scrollIndicatorInsets.bottom = bottomInset
+        let bottomInset = calculateBottomInset(with: keyboardEndFrame)
+        let shouldUpdateBottomInset = bottomInset != messagesCollectionView.contentInset.bottom
 
-        } else {
-            //Software keyboard is found
-            let bottomInset = keyboardEndFrame.height > keyboardOffsetFrame.height ? (keyboardEndFrame.height - iPhoneXBottomInset) : keyboardOffsetFrame.height
-            messagesCollectionView.contentInset.bottom = bottomInset
-            messagesCollectionView.scrollIndicatorInsets.bottom = bottomInset
+        guard shouldUpdateBottomInset else { return }
+
+        UIView.animate(withDuration: 0.5) {
+            self.adjustBottomContentInset(bottomInset ?? 0)
```

Comment:
Does this need to be in an animation block?

Reply:
Yes, we have to use animation here, otherwise it'll just snap place.
I tried to mimick iMessage behavior here.

Should MessageKit mimick iMessage behavior? This is a question :)